packages/perspective-jupyterlab/test/jupyter/utils.js

```diff
@@ -112,17 +121,49 @@ module.exports = {
         // wait for a cell to be active
         await page.waitForSelector(
             '.jp-Notebook-ExecutionIndicator:not([data-status="idle"])',
-            {timeout: 3000}
+            {timeout: 5000}
         );
 
         await page.waitForSelector(
             '.jp-Notebook-ExecutionIndicator[data-status="idle"]',
-            {timeout: 3000}
+            {timeout: 10000}
         );
 
         // Use our custom keyboard shortcut to run all cells
         await page.keyboard.press("R");
         await page.keyboard.press("R");
         await page.evaluate(() => (document.scrollTop = 0));
     },
+    add_and_execute_cell: async (page, cell_content) => {
+        await page.waitForSelector(".jp-CodeCell", {
+            visible: true,
+        });
+
+        // find and click the a cell in the notebook
+        await page.click(".jp-CodeCell");
+
+        // find and click the "new cell" button
+        await page.click(
+            '.jp-Button[data-command="notebook:insert-cell-below"]'
+        );
+
+        // after clicking new cell, the document will auto
+        // focus the new cell, so lets grab it
+        const el = await page.evaluateHandle(() => document.activeElement);
+        await el.type(cell_content);
+
+        // now while the element is still focused, click the run cell button
+        await page.click('.jp-Button[data-command="runmenu:run"]');
+
+        // wait a small time
+        await new Promise((res) => setTimeout(res, 2000));
```

Comment:
Ditto - anything that involves "wait an arbitrary time" is pretty much guaranteed to fail eventually.  Certainly there is an appropriate condition to await here ...

Reply:
There is no guaranteed condition to wait for here, this code is running arbitrary python code that may or may not mutate any frontend state. If you're not mutating frontend state, you won't have anything for puppeteer to wait on and in my experimentation the idle kernel state indicator is also unreliable.

Reply:
You can await on the presence of the DOM structure output from the cell, the `Out[#]` cell indicator, or a number of other DOM changes.   `PerspectiveWidget` tests in this file do so with `.jp-OutputArea-output perspective-viewer`.  You may need to use interactive tests to check what the DOM looks like at resolution - but if there truly is "no guaranteed condition to wait for" ... then there is nothing to test.

Reply:
How long do you wait before timing out?

Reply:
Also which cell's `Out[#]` do you wait on? How do you determine one cell's output from another preexisting cells output? Can you provide an example of puppeteer waiting for specific text content?